README.markdown
# darksky

Ruby gem for retrieving data from the [Dark Sky API](http://darkskyapp.com/api/). The Dark Sky API lets you query
for short-term precipitation forecast data at geographical points inside the United States.

## Installation

`gem install darksky`

or in your `Gemfile`

```ruby
gem 'darksky'
```

## Usage

```ruby
darksky = Darksky::API.new('this-is-your-dark-sky-api-key')

# Returns a forecast for the next hour at a given location.
forecast = darksky.forecast('42.7243','-73.6927')

# Returns a brief forecast for the next hour at a given location.
brief_forecast = darksky.brief_forecast('42.7243','-73.6927')

# Returns forecasts for a collection of arbitrary points.
precipitation = darksky.precipitation(['42.7','-73.6',1325607100,'42.0','-73.0',1325607791])

# Returns a list of interesting storms happening right now.
interesting_storms = darksky.interesting

# Returns historic data for a given location.
history = darksky.time_machine('42.7243','-73.6927', '409467600')
```

## Contributing to darksky

* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it
* Fork the project
* Start a feature/bugfix branch
* Commit and push until you are happy with your contribution
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
* Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

## Copyright

Copyright (c) 2012-2014 David Czarnecki. See LICENSE.txt for further details.


CHANGELOG.markdown
# CHANGELOG

## 1.0.6
* Fixed API url

## 1.0.5

* Updated gemspec to reflect new homepage. The repository has moved to https://github.com/darkskyapp/darksky-ruby.

## 1.0.4

* Added `brief_forecast` API endpoint to return a brief forecast for the latitude and longitude.

## 1.0.3

* Added `interesting` API endpoint to return a list of interesting storms happening right now.
* Fixed bug in `precipitation` API endpoint where multiple locations were not correctly passed to the API.

## 1.0.2

* Removed default to disable SSL peer verification now that the DarkSky API server has a signed certificate.

## 1.0.1

* Added `options` hash to API endpoints as final parameter.
* Default options for API endpoints disable SSL peer verification while DarkSky API is in beta and still uses a self-signed certificate.
* Fixed response parsing for `precipitation` API endpoint.

## 1.0.0

* Initial release.


spec/spec_helper.rb
require 'rspec'
require 'darksky'
require 'vcr'

Dir["#{File.dirname(__FILE__)}/support/**/*.rb"].each {|f| require f}

VCR.configure do |c|
  c.cassette_library_dir = 'spec/cassettes'
  c.hook_into :typhoeus
  c.allow_http_connections_when_no_cassette = false
end

RSpec.configure do |config|
end


spec/darksky/api_spec.rb
require 'spec_helper'

describe Darksky::API do
  let(:darksky_api) { Darksky::API.new('this-is-your-dark-sky-api-key') }

  describe '#forecast' do
    it 'should return a valid forecast for a latitude and longitude' do    
      VCR.use_cassette('forecast', :record => :once) do
        forecast = darksky_api.forecast('42.7243','-73.6927')
        forecast['currentSummary'].should == 'clear'
        forecast['radarStation'].should == 'enx'
      end
    end
  end

  describe '#brief_forecast' do
    it 'should return a valid brief forecast for a latitude and longitude' do    
      VCR.use_cassette('brief_forecast', :record => :once) do
        forecast = darksky_api.brief_forecast('42.7243','-73.6927')
        forecast['currentSummary'].should == 'clear'
        forecast['hourSummary'].should == 'clear'
      end
    end
  end

  describe '#precipitation' do
    it 'should return forecasts for a collection of arbitrary points' do
      VCR.use_cassette('precipitation', :record => :once) do
        precipitation = darksky_api.precipitation(['42.7','-73.6',1325607100,'42.0','-73.0',1325607791])
        precipitation['precipitation'].size.should == 2
        precipitation['precipitation'].first['probability'].should == 0
        precipitation['precipitation'].first['type'].should == 'rain'
      end
    end
  end

  describe '#interesting' do
    it 'should return a list of interesting storms happening right now' do
      VCR.use_cassette('interesting', :record => :once) do
        interesting_storms = darksky_api.interesting
        interesting_storms.size.should == 1
      end
    end
  end

  describe '#time_machine' do
    it 'should return a valid forecast for a latitude and longitude' do    
      VCR.use_cassette('forecast', :record => :once) do
        forecast = darksky_api.time_machine('42.7243','-73.6927','409467600')
        forecast['currentSummary'].should == 'clear'
        forecast['radarStation'].should == 'enx'
      end
    end
  end
end

spec/darksky/version_spec.rb
require 'spec_helper'

describe 'Darksky::VERSION' do
  it 'should be the correct version' do
    Darksky::VERSION.should == '1.0.5'
  end
end

lib/darksky/api.rb
require 'typhoeus'
require 'json'

module Darksky
  class API
    DARKSKY_API_URL = 'https://api.darksky.net'
    DEFAULT_OPTIONS = {
    }

    # Create a new instance of the Darksky::API using your API key.
    #
    # @param api_key [String] Dark Sky API key.
    def initialize(api_key)
      @api_key = api_key
    end

    # Returns a forecast for the next hour at a given location.
    #
    # @param latitude [String] Latitude in decimal degrees.
    # @param longitude [String] Longitude in decimal degrees.
    # @param option [Hash] (Optional) Options to be passed to the Typhoeus::Request
    def forecast(latitude, longitude, options = {})
      response = Typhoeus::Request.get("#{DARKSKY_API_URL}/forecast/#{@api_key}/#{latitude},#{longitude}", DEFAULT_OPTIONS.dup.merge(options))
      JSON.parse(response.body) if response.code == 200
    end

    # Returns a brief forecast for the next hour at a given location.
    #
    # @param latitude [String] Latitude in decimal degrees.
    # @param longitude [String] Longitude in decimal degrees.
    # @param option [Hash] (Optional) Options to be passed to the Typhoeus::Request
    def brief_forecast(latitude, longitude, options = {})
      response = Typhoeus::Request.get("#{DARKSKY_API_URL}/brief_forecast/#{@api_key}/#{latitude},#{longitude}", DEFAULT_OPTIONS.dup.merge(options))
      JSON.parse(response.body) if response.code == 200
    end

    # Returns forecasts for a collection of arbitrary points.
    #
    # @param latitudes_longitudes_times [Array] Triplets of latitude, longitude and time. Example: ['42.7','-73.6',1325607100,'42.0','-73.0',1325607791]
    # @param option [Hash] (Optional) Options to be passed to the Typhoeus::Request
    def precipitation(latitudes_longitudes_times, options = {})
      return if latitudes_longitudes_times.size % 3 != 0
      params = []
      latitudes_longitudes_times.each_slice(3) do |data|
        params << data.join(',')
      end
      response = Typhoeus::Request.get("#{DARKSKY_API_URL}/precipitation/#{@api_key}/#{params.join(';')}", DEFAULT_OPTIONS.dup.merge(options))
      JSON.parse(response.body) if response.code == 200
    end

    # Returns a list of interesting storms happening right now.
    #
    # @param option [Hash] (Optional) Options to be passed to the Typhoeus::Request
    def interesting(options = {})
      response = Typhoeus::Request.get("#{DARKSKY_API_URL}/interesting/#{@api_key}", DEFAULT_OPTIONS.dup.merge(options))
      JSON.parse(response.body) if response.code == 200
    end
  end
end
